### PR TITLE
Adds Redis Monitoring

### PR DESCRIPTION
**Pull request by sapiderman**
# Monika Pull Request (PR)  

## What feature/issue does this PR add  
1. Closing issue #858 
2. Add new probe for redis

## How did you implement / how did you fix it  
1.  Using redis lib
2. Adding redis as a "request type"
3. Add to documentation

## How to test  
1. Use the following config `monika.redis.yaml`:
```yaml
probes:
  - id: checking-redis
    name: localhost all works
    alerts:
      - query: response.status == 500
        message: response status message
    incidentThreshold: 1
    recoveryThreshold: 1
    redis:
      - host: 172.17.0.1
        port: 6379
      - host: 172.17.0.2
        port: 6379

  - id: checking-bad-redis
    name: bad redis check
    alerts:
      - query: response.status == 500
        message: response status message
    incidentThreshold: 1
    recoveryThreshold: 1
    redis:
      - host: 172.17.0.10
        port: 6379
        password: mypassword

notifications:
  - id: desktop-01
    type: desktop
```

2. Run your monika with `-c monika.redis.yaml`
4. You should see something like:

image

## Screenshot

Notification
image

Alert and Notification
image

Use of password with and without if Redis has AUTH set:
image

### CODE REVIEW

File: src/components/redis-request/index.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+async function sendRedisRequest(params: RedisRequest): Promise<RedisResult> {
+  const { host, port } = params
+
+  const result: RedisResult = {
+    isAlive: false,
+    message: '',
+  }
+  try {
+    const client = createClient({
+      // redis[s]://[[username][:password]@][host][:port][/db-number]:
```
Comment (dennypradipta) [inline src/components/redis-request/index.ts:111]: Remove unused comment

File: src/components/redis-request/index.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
+/**********************************************************************************
```
Comment (dennypradipta) [inline src/components/redis-request/index.ts:1]: Remove doubled license

File: src/components/probe/index.ts [hunk 3]
```diff
@@ -151,6 +152,63 @@ export async function doProbe({
...
   const responses = []
 
+  if (probe?.redis) {
+    const { id, redis } = probe
+
+    let redisRequestIndex = 0
+    for (const redisIndex of redis) {
+      const { host, port } = redisIndex
+
+      // eslint-disable-next-line no-await-in-loop
```
Comment (dennypradipta) [inline src/components/probe/index.ts:162]: If you want to use await in loop, you can change the line 159 to:

```
for await (const redisIndex of redis) {
   ...
}
```

**Review by dennypradipta — APPROVED** [review]
Approved with suggestions. Good work!
image

**Review by dennypradipta — APPROVED** [review]

File: docs/src/pages/guides/probes.md [hunk 4]
```diff
@@ -89,6 +89,21 @@ probes:
...
+
+```yaml
+probes:
+  - id: 'redis-test'
+    name: redis health
+    description: requesting redis PONG
+    interval: 30 # in seconds
+    redis:
+      - host: 172.15.0.2
+        port: 6379
```
Comment (nicnocquee) [inline docs/src/pages/guides/probes.md:104]: Add `password` too

**Review by nicnocquee — CHANGES_REQUESTED** [review]

File: docs/src/pages/guides/probes.md [hunk 5]
```diff
@@ -89,6 +89,38 @@ probes:
         ping: true
 ```
 
+### Redis Request
+
+You can query your redis instance. The following probe sends a ping to your redis host every 30 seconds.
```
Comment (nicnocquee) [inline docs/src/pages/guides/probes.md:94]: You can check if your redis instance is running and accessible by adding a probe with `redis` configuration as follows.

File: docs/src/pages/guides/probes.md [hunk 6]
```diff
@@ -89,6 +89,38 @@ probes:
...
+    name: redis check with password
+    description: requesting redis PONG
+    interval: 30 # in seconds
+    redis:
+      - host: 172.15.0.2
+        port: 6379
+        password: mypassword
+```
+
+The example above use only password, but both `username` and `password` are possible and optional.
```
Comment (nicnocquee) [inline docs/src/pages/guides/probes.md:122]: You may also add `username` property if needed.

File: src/components/config/validate.ts [hunk 7]
```diff
@@ -435,3 +445,22 @@ function validateTCPConfig(tcpConfig?: Socket) {
...
 }
+
+function validateRedisConfig(redisConfig?: Redis[]) {
+  if (!redisConfig) {
+    return ''
+  }
+
+  const schema = Joi.object({
+    host: Joi.string().required(),
+    port: Joi.number().required(),
```
Comment (nicnocquee) [inline src/components/config/validate.ts:456]: maybe we can validate the port here, it should be in the range of 0-65536, right? 

```
Joi.number().integer().min(0).max(65536)
```

**Review by nicnocquee — CHANGES_REQUESTED** [review]

File: src/components/config/validate.ts [hunk 8]
```diff
@@ -435,3 +445,22 @@ function validateTCPConfig(tcpConfig?: Socket) {
...
   return validationError?.error?.message
 }
+
+function validateRedisConfig(redisConfig?: Redis[]) {
+  if (!redisConfig) {
+    return ''
+  }
+
+  const schema = Joi.object({
+    host: Joi.string().required(),
```
Comment (nicnocquee) [inline src/components/config/validate.ts:455]: Wonder if we can validate `host` to be hostname or ip address like this 

```
    host: Joi.valid([Joi.string().hostname(), Joi.string().ip()]).required(),

```

File: src/components/redis-request/index.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+  const result: RedisResult = {
+    isAlive: false,
+    message: '',
+  }
+
+  try {
+    const client = createClient({
+      socket: {
+        host: '172.17.0.1',
```
Comment (nicnocquee) [inline src/components/redis-request/index.ts:91]: is this correct? why not use the `host` variable

**Review by nicnocquee — CHANGES_REQUESTED** [review]

File: src/components/config/validate.ts (hunk 8 above)
Comment (sapiderman) [reply src/components/config/validate.ts:455]: TIL 😁 
`host: Joi.alternatives().try(Joi.string().hostname(), Joi.string().ip()).required()`

**Review by nicnocquee — APPROVED** [review]

**Pull request merged**